### PR TITLE
Fixes #48 - Better handling of background images

### LINKED ISSUE #48: Better handling of background images
Background images in grid elements are not mobile optimized and allows images of any size to be inserted.

Example: When using _Advanced one column grid_ and adding a background image, the original image is used directly:
`style="background-image: url('/fileadmin/user_upload/myverylargeimage.jpg')"`

This opens up for a couple of issues:
1. Editors can upload very large images and site load page times will be increased.
2. Images are not optimized for smaller screens. An image 3000px wide is not necessary for mobile devices and will contribute to high load times on slow mobile connections.

Suggested solution: Providing resized images with CSS media queries. 

Example implementation: Here's a TypoScript implementation using the page:media field. It should be possible to adapt:

```
page.cssInline {
    20 = FILES
    20 {

        references {
            table = pages
            uid.data = TSFE:id
            fieldName = media
        }

        renderObj = COA
        renderObj {
            5 = LOAD_REGISTER
            5 {
                mediaImageIndex = TEXT
                mediaImageIndex {
                    cObject = TEXT
                    cObject {
                        data = register:sectionIteration
                        intval = 1
                        outerWrap = |+1
                    }

                    prioriCalc = 1
                }
            }

            10 = TEXT
            10 {
                prepend = TEXT
                prepend {
                    data = register:mediaImageIndex
                    noTrimWrap = |.media-image-| { background-image: url(/|
                }

                cObject = IMG_RESOURCE
                cObject {
                    file {
                        import.data = file:current:publicUrl
                        width = 1170
                    }
                }

                append = TEXT
                append {
                    value = ); }
                }
            }

            20 = TEXT
            20.cObject < .10
            20 {
                cObject.cObject.file.width = 970
                noTrimWrap = | @media screen and (max-width: 970px) { | } |
            }

            30 = TEXT
            30.cObject < .10
            30 {
                cObject.cObject.file.width = 750
                noTrimWrap = | @media screen and (max-width: 970px) { | } |
            }

            40 = TEXT
            40.cObject < .10
            40 {
                cObject.cObject.file.width = 480
                noTrimWrap = | @media screen and (max-width: 970px) { | } |
            }
        }
    }
}
```

### PR DESCRIPTION
**Pull request by mabolek**

### CODE REVIEW

File: Configuration/GridElements/TypoScript/lib.responsiveBackgroundImage.setupts [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+            current = 1
+            trim = 1
+        }
+    }
+
+}
+
+page.cssInline {
+    # Debug related output
```
Comment (MattiasNilsson) [inline Configuration/GridElements/TypoScript/lib.responsiveBackgroundImage.setupts]: Remove commented code.

**Review by MattiasNilsson — CHANGES_REQUESTED** [review]
This looks good to me, just some smaller nitpicks! 👍

**Comment by pixelmatseriks** [comment]
It is a great feature to be able to have different images for different devices, but this commit will affect all GridElement CE:s (f.ex. Parallax), not just the advanced grid. We need to have a way to enable it per CE. Maybe there is a way to implement it using the new "cropVariant"?

**Comment by mabolek** [comment]
@pixelmatseriks cropVariant support would probably be good in any case (I haven't looked too closely at it yet), but regarding the feature affecting other grid element CEs: Wouldn't it be good to have responsive images for parallax as well? Have I missed your point? 🙂

**Comment by pixelmatseriks** [comment]
@mabolek Not all of the gridelements CE:s uses the image as a background image.

**Comment by mabolek** [comment]
@pixelmatseriks Hmm... In that case, the current implementation has the same problem. All GridElements have a `style="background-image: url('{f:uri.image(src: '{files.0.uid}', treatIdAsReference: 1)}')"` if there's a file referenced in the content element. Only difference is that this pull request compresses the file so it's no larger than necessary.

**Comment by pixelmatseriks** [comment]
@mabolek hmm, you are right. The Default layout always include a background-image if there's a file referenced in the content element, but the Empty layout doesn't. There is a condition in Collapsible, SimpleAccordion and Tab to use Default or Empty, but all of the other templates uses Default. The Parallax also includes file.0 if present in the template itself, but maybe it is a "fallback".

**Comment by mabolek** [comment]
@pixelmatseriks I suggest we use the current implementation and open separate feature requests for 1) implementing cropVariant and 2) improving the way layouts work, so files are only included in CEs that need them.

**Comment by dmh** [comment]
#211

**Pull request closed without merging**